### ISSUE TITLE
Allow user-specified step match names

### ISSUE THREAD

**Issue by timriley** [dry-rb/dry-transaction#34]
_Copying @mrbongiolo's comments over from https://github.com/dry-rb/dry-transaction/pull/33_

<hr>

Do you still want to leave the `step_name` as default matching behavior? Couldn't we also provide a more general approach, in which we define the kind of failure type with an option `failure: :something` on the transactions? I would happily send a PR for this.

What about having: `ResultMatcher = StepNameMatcher` or `ResultMatcher = FailureOptionMatcher`, so the library would still call `ResultMatcher` but I could have a completely different behavior for an app.

I have transactions defined like this:

``` ruby
perform do
  tee_validate :resource_status_validation, call_with_args: true
  validate :schema
  tee :update_resource
  tee :notify_professional
  map :decorate
end
```

In this case, both steps `:schema` and `:resource_status_validation` should have the same kind of `:unprocessable_entity` failure, right now I'm raising a custom error to do it, but having the ability to define a `:failure` option on the step would make it much better.

So I would have:

``` ruby
perform do
  tee_validate :resource_status_validation, call_with_args: true, failure: :unprocessable_entity
  validate :schema, failure: :unprocessable_entity
  tee :update_resource, failure: :some_weird_failure?
  tee :notify_professional
  map :decorate
end
```

This would require a small change to some StepAdapters. This might not be an approach needed by everyone, and that's fine, but it appeared in more than one place on my use cases.

**Comment by mrbongiolo** [comment]
@timriley could we pass the step options or even the entire step into the `StepFailure`? This would allow me to build the custom matcher using the `failure:` option.
After all it makes sense that the `StepFailure` knows about the step that failed.

In this way the match: `result.value` would have access to more informations about that step.

**Comment by timriley** [comment]
This is fixed now courtesy of #35 :)
